src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -751,114 +741,219 @@ private void ReturnReadBufferIfEmpty()
-                        // We don't have enough bytes buffered, so issue an initial read to try to get enough.  This is
-                        // done in this method both to better consolidate error handling logic (the first read is the special
-                        // case that needs to differentiate reading 0 from > 0, and everything else needs to throw if it
-                        // doesn't read enough), and to minimize the chances that in the common case the FillBufferAsync
-                        // helper needs to yield and allocate a state machine.
-                        int readBytes = await adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount)).ConfigureAwait(false);
-                        if (readBytes == 0)
+                        if (_internalBufferCount != 0)
                         {
-                            return 0;
+                            // we got EOF in middle of TLS frame. Treat that as error.
+                            throw new IOException(SR.net_io_eof);
                         }
 
-                        _internalBufferCount += readBytes;
-                        if (_internalBufferCount < SecureChannel.ReadHeaderSize)
-                        {
-                            await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize).ConfigureAwait(false);
-                        }
+                        return 0;
                     }
-                    Debug.Assert(_internalBufferCount >= SecureChannel.ReadHeaderSize);
 
-                    // Parse the frame header to determine the payload size (which includes the header size).
-                    int payloadBytes = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
-                    if (payloadBytes < 0)
+                    _internalBufferCount += bytesRead;
+                    if (frameSize == int.MaxValue && _internalBufferCount > SecureChannel.ReadHeaderSize)
                     {
-                        throw new IOException(SR.net_frame_read_size);
+                        frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
                     }
 
-                    // Read in the rest of the payload if we don't have it.
-                    if (_internalBufferCount < payloadBytes)
+                    if (_internalBufferCount < frameSize)
                     {
-                        await FillBufferAsync(adapter, payloadBytes).ConfigureAwait(false);
+                        t = adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount));
+                        continue;
                     }
 
-                    // Set _decrytpedBytesOffset/Count to the current frame we have (including header)
-                    // DecryptData will decrypt in-place and modify these to point to the actual decrypted data, which may be smaller.
-                    _decryptedBytesOffset = _internalOffset;
-                    _decryptedBytesCount = payloadBytes;
+                    break;
```

Comment:
It's kind of weird to have a while(true) with a break at the end. Should this just be a do/while loop?

Reply:
I updated the code. It will check _internalBufferCount twice but maybe compiler can optimize and it would be more readable.